docs/DafnyRef/GrammarDetails.md

```diff
@@ -0,0 +1,1685 @@
+manual](http://www.ssw.uni-linz.ac.at/Research/Projects/Coco/Doc/UserManual.pdf)
+for details.
+The Dafny input file to Coco/R is the `Dafny.atg` file in the source tree.
+
+The grammar is an _attributed extended BNF_ grammar.
+The _attributed_ adjective indicates that the BNF productions are
+parameterized by boolean parameters that control variations of the 
+production rules, such as whether a particular alternative is permitted or
+not. Using such attributes allows combining non-terminals with quite
+similar production rules, making a simpler, more compact and more
+readable grammer.
+
+The grammar rules presented here replicate those in the source
+code, but omit semantic actions, error recovery markers, and
+conflict resolution syntax. Some uses of the attribute
+parameters are described informally.
+
+The names of character sets and tokens start with a lower case
+letter; the names of grammar non-terminals start with
+an upper-case letter.
+
+
+
+
+## 29.1. Dafny Syntax
+
+This section gives the definitions of Dafny tokens.
+
+### 29.1.1. Classes of characters
+
+These definitions define some names as representing subsets of the set of characters. Here
+
+* double quotes enclose the set of characters constituting the class, 
+* single quotes enclose a single character (perhaps an escaped representation using `\`), 
+* the binary `+` indicates set union, 
+* binary `-` indicates set difference, and 
+* `ANY` indicates the set of all (unicode) characters.
+
+````grammar
+letter = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
+
+digit = "0123456789"
+posDigit = "123456789"
+posDigitFrom2 = "23456789"
+
+hexdigit = "0123456789ABCDEFabcdef"
+
+special = "'_?"
+
+cr        = '\r'
+
+lf        = '\n'
+
+tab       = '\t'
+
+space     = ' '
+
+nondigitIdChar = letter + special
+
+idchar = nondigitIdChar + digit
+
+nonidchar = ANY - idchar
+
+charChar = ANY - '\'' - '\\' - cr - lf
+
+stringChar = ANY - '"' - '\\' - cr - lf
+
+verbatimStringChar = ANY - '"'
+````
+
+A `nonidchar` is any character except those that can be used in an identifier.
+Here the scanner generator will interpret `ANY` as any unicode character.
+However, `nonidchar` is used only to mark the end of the `!in` token;
+in this context any character other than [whitespace or printable ASCII](#sec-unicode)
+will trigger a subsequent scanning or parsing error.
+
+### 29.1.2. Definitions of tokens {#sec-g-tokens}
+
+These definitions use 
+
+* double-quotes to indicate a verbatim string (with no escaping of characters)
+* `'"'` to indicate a literal double-quote character
+* vertical bar to indicate alternatives
+* square brackets to indicate an optional part
+* curly braces to indicate 0-or-more repetitions
+* parentheses to indicate grouping
+* a `-` sign to indicate set difference: any character sequence matched by the left operand except character sequences matched by the right operand
+* a sequence of any of the above indicates concatentation (without whitespace)
```

Comment:
The form of this last bullet is different from those before it.

Reply:
Thanks - fixed.